Add template documentation (#960)
nautobot/nautobot: nautobot/docs/development/templates.md

> # Templates for Plugins

Reviewer: ```suggestion
# Page Templates
```

> Nautobot comes with a variety of plugins that allow for a lot of flexibility

Reviewer: ```suggestion
Nautobot comes with a variety of page templates that allow for a lot of flexibility
```

while keeping the page style consistent with the rest of the application.
This document presents these templates and their features.

> You can use it for your templates by calling `{% extends '<template_name>' %}`

Reviewer: ```suggestion
You can use these templates as the basis for your templates by calling `{% extends '<template_name>' %}`
```

at the top of your template file.

## Object Detail

> The most customizable template available to plugin authors is
> `generic/object_detail.html`.
> 
> The following blocks are available to you:

Reviewer: ```suggestion
The most customizable template is `generic/object_detail.html`, as object detail views have a wide range of specific requirements to be accommodated. It provides the following blocks:
```

      section on the right of the page.
        - `extra_buttons`: this block enables extending the buttons block
          without losing the predefined buttons. Custom buttons will appear
          between the plugin buttons and clone/edit/delete actions.
    - `masthead`: is the block that contains the title. Overloading it enables
      to change anything about the title block.
    - `title`: is the block contained by `masthead` and wrapped in a heading
      block. Overloading it makes it possible to change the heading text as
      well as the page title shown in the browser.
    - `nav_tabs`: are the navigation tabs. If overloaded, custom tabs can be
      rendered instead of the default.
        - `extra_nav_tabs`: this block allows to add new tabs without having to
          override the default ones.
- `content`: is the entire content of the page below the `header`.
    - `content_left_page`: is a half-width column on the left. Multiple panels
      can be rendered in a single block.
    - `content_right_page`: is a half-width column on the right.
    - `content_full_width_page`: is a full-width column.

## Object List

> The base view for listing objects is `generic/object_list.html`.
> 
> In it, the following blocks may be implemented:

Reviewer: ```suggestion
The base template for listing objects is `generic/object_list.html`, with the following blocks:
```

- `buttons`: may provide a set of buttons at the top right of the page, to the
  left of the table configuration button.
- `sidebar`: may implement a sidebar below the search form on the right.
- `bulk_buttons`: may be a set of buttons at the bottom of the table, to the
  left of potential bulk edit or delete buttons.

## Object Edit

> The base view for object addition or change is `generic/object_edit.html`. It

Reviewer: ```suggestion
The base template for object addition or change is `generic/object_edit.html`. It
```

> does not provide any blocks for customizing the user experience.

Reviewer: Actually many models override the `block form` in this template to customize the form rendering and layout beyond what a generic `{% render_form form %}` provides - probably worth mentioning.

## Object Import

> The base view for object import is `generic/object_import.html`.
> 
> The blocks that views may override are:

Reviewer: ```suggestion
The base template for object import is `generic/object_import.html`, with the following blocks:
```

- `tabs`: may provide tabs at the top of the page. The default import view is
  not tabbed.

## Object Deletion

> The base view for object deletion is `generic/object_delete.html`.
> 
> It provides the following custom blocks:

Reviewer: ```suggestion
The base template for object deletion is `generic/object_delete.html`, with the following blocks:
```

- `message`: is the overridable confirmation message for deletion.
    - `message_extra`: provides a way to add to the default message without
      overriding it.

## Bulk Edit

> The base view for bulk object change is `generic/object_bulk_edit.html`. It

Reviewer: ```suggestion
The base template for bulk object change is `generic/object_bulk_edit.html`. It
```

does not provide any blocks for customizing the user experience.

## Bulk Import

> The base view for bulk object import is `generic/object_bulk_import.html`.
> 
> The blocks that views may override are:

Reviewer: ```suggestion
The base template for bulk object import is `generic/object_bulk_import.html`, with the following blocks:
```

- `tabs`: may provide tabs at the top of the page. The default import view is
  not tabbed.

## Bulk Deletion

> The base view for bulk object deletion is `generic/object_bulk_delete.html`.
> 
> It provides the following custom blocks:

Reviewer: ```suggestion
The base template for bulk object deletion is `generic/object_bulk_delete.html`, with the following blocks:
```

- `message_extra`: provides a way to add to the default message.

> **Note**: contrary to the deletion of a single object, this view does *not*

Reviewer: ```suggestion
**Note**: contrary to the deletion of a single object, this template does *not*
```

provide a way to completely override the deletion message.

## Bulk Renaming

> The base view for renaming object in bulk is `generic/object_bulk_rename.html`.

Reviewer: ```suggestion
The base template for renaming objects in bulk is `generic/object_bulk_rename.html`.
```
